SkEditor/Utilities/InternalAPI/AddonLoader.cs

```diff
@@ -0,0 +1,342 @@
+            return;
+        }
+        
+        if (Addons.Any(m => m.Addon.Identifier == addon[0].Identifier))
+        {
+            SkEditorAPI.Logs.Warning($"Failed to load addon from \"{dllFile}\": An addon with the identifier \"{addon[0].Identifier}\" is already loaded.");
+            return;
+        }
+        
+        Addons.Add(new AddonMeta()
+        {
+            Addon = addon[0],
+            State = IAddons.AddonState.Installed,
+            DllFilePath = dllFile,
+            Errors = [],
+            LoadContext = loadContext
+        });
+
+        await EnableAddon(addon[0]);
+    }
+
+    public static async void LoadAddon(Type addonClass)
+    {
+        var addon = (IAddon) Activator.CreateInstance(addonClass);
+        Addons.Add(new AddonMeta()
+        {
+            Addon = addon, State = IAddons.AddonState.Disabled,
+            Errors = []
+        });
+
+        await EnableAddon(addon);
+    }
+
+    private static bool CanEnable(AddonMeta addonMeta)
+    {
+        var addon = addonMeta.Addon;
+        var minimalVersion = addon.GetMinimalSkEditorVersion();
+        if (SkEditorAPI.Core.GetAppVersion() < minimalVersion)
+        {
+            SkEditorAPI.Logs.Debug($"Addon \"{addon.Name}\" requires SkEditor version {minimalVersion}, but the current version is {SkEditorAPI.Core.GetAppVersion()}. Disabling it.");
+            addonMeta.State = IAddons.AddonState.Disabled;
+            addonMeta.Errors.Add(LoadingErrors.OutdatedSkEditor(minimalVersion));
+            return false;
+        }
+                
+        var maximalVersion = addon.GetMaximalSkEditorVersion();
+        if (maximalVersion != null && SkEditorAPI.Core.GetAppVersion().CompareTo(maximalVersion) > 0)
+        {
+            SkEditorAPI.Logs.Debug($"Addon \"{addon.Name}\" requires SkEditor version {maximalVersion}, but the current version is {SkEditorAPI.Core.GetAppVersion()}. Disabling it.");
+            addonMeta.State = IAddons.AddonState.Disabled;
+            addonMeta.Errors.Add(LoadingErrors.OutdatedAddon(maximalVersion));
+            return false;
+        }
+        
+        return true;
+    }
+
+    /// <summary>
+    /// Preload an addon. This will make the desired addon ready to be enabled,
+    /// by checking if it can actually be enabled, and its dependencies.
+    /// </summary>
+    /// <param name="addonMeta">The addon to preload.</param>
+    /// <returns>True if the addon can be enabled, false otherwise.</returns>
+    public static async Task<bool> PreLoad(AddonMeta addonMeta)
+    {
+        if (!CanEnable(addonMeta))
+            return false;
+
+        if (!await LocalDependencyManager.CheckAddonDependencies(addonMeta))
+            return false;
+
+        if (addonMeta.NeedsRestart)
+        {
+            await SkEditorAPI.Windows.ShowMessage("Addon needs restart", $"The addon \"{addonMeta.Addon.Name}\" needs a restart to be enabled correctly.");
```

Comment:
This shouldn't execute if the addon is loaded during the app startup (instead of being installed from the Marketplace). When I tested Analyzer it was behaving like this.

Reply:
I don't know how I can make it work, as the addon must be loaded one time (to gather meta-like dependencies) then another time (when the dependencies are downloaded), while being unloaded between those two steps. As the "context" changes (e.g. the assembly loader has to take care of the libs for the second step) I can't do that on runtime :/